File: domain/goal.ts
export type GoalType = "yearly" | "quarterly" | "monthly" | "weekly";

export type GoalStatus = "active" | "completed" | "paused";

export interface Goal {
  id: string;
  userId: string;
  title: string;
  description?: string;
  type: GoalType;
  // Period identification
  year: number; // Always required
  quarter?: number; // For quarterly goals (1-4)
  month?: number; // For monthly goals (1-12)
  week?: number; // For weekly goals (ISO week number)
  // Progress tracking
  targetValue: number; // Target value (e.g., 10 books, 10000 steps)
  currentValue: number; // Current progress
  unit: string; // Unit of measurement (e.g., "books", "hours", "steps")
  // Task linking (optional)
  linkedTaskIds: string[]; // IDs of tasks that contribute to this goal
  // Status and metadata
  status: GoalStatus;
  createdAt: string;
  updatedAt: string;
  completedAt?: string;
  // Progress calculation method
  progressType: "manual" | "task-linked"; // How progress is calculated
}

File: backend/app/schemas/goal.py
from pydantic import BaseModel
from typing import Optional, List
from datetime import datetime

class GoalBase(BaseModel):
    title: str
    description: Optional[str] = None
    type: str  # "yearly", "quarterly", "monthly", "weekly"
    year: int
    quarter: Optional[int] = None
    month: Optional[int] = None
    week: Optional[int] = None
    targetValue: float
    currentValue: float = 0.0
    unit: str
    linkedTaskIds: Optional[str] = None  # JSON array
    status: str = "active"
    progressType: str = "manual"

class GoalCreate(GoalBase):
    id: str
    userId: str

class Goal(GoalBase):
    id: str
    userId: str
    createdAt: datetime
    updatedAt: datetime
    completedAt: Optional[datetime] = None

    class Config:
        from_attributes = True
